src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -302,11 +303,42 @@ private Task WriteSingleChunk<TWriteAdapter>(TWriteAdapter writeAdapter, ReadOnl
                 return WaitForWriteIOSlot(writeAdapter, ioSlot, buffer);
             }
 
-            byte[] rentedBuffer = ArrayPool<byte>.Shared.Rent(buffer.Length + FrameOverhead);
-            byte[] outBuffer = rentedBuffer;
+            Task semaphoreTask = writeAdapter.ThrottleAsync();
 
-            SecurityStatusPal status = _sslState.EncryptData(buffer, ref outBuffer, out int encryptedBytes);
+            if (semaphoreTask.IsCompletedSuccessfully)
+            {
+                byte[] rentedBuffer = ArrayPool<byte>.Shared.Rent(buffer.Length + FrameOverhead);
+                byte[] outBuffer = rentedBuffer;
+
+                SecurityStatusPal status = _sslState.EncryptData(buffer, ref outBuffer, out int encryptedBytes);
+                s_throttle.Release();
+                return WriteEncryptedDataAsync(writeAdapter, outBuffer, rentedBuffer, encryptedBytes, status);
```

Comment:
Can these five lines be moved to a Task-returning method?  Then it's essentially:
```C#
return semaphoreTask.IsCompletedSuccessfully ?
    ThatMethodWhateverItsCalled() :
    CompleteAsync(writeAdapter, semaphoreTask);

async Task CompleteAsync(TWriteAdapter wAdapter, Task sTask)
{
    await sTask.ConfigureAwait(false);
    await ThatMethodWhateverItsCalled().ConfigureAwait(false);
}
```